File: android/app/src/main/java/com/modi/connect/core/connectivity/session/LegacySessionObserver.kt
package com.modi.connect.core.connectivity.session

import com.modi.connect.ConnectionState
import com.modi.connect.core.connectivity.SessionId
import com.modi.connect.core.connectivity.SessionObservationResult
import com.modi.connect.core.connectivity.SessionRegistry
import com.modi.connect.core.connectivity.SessionState
import com.modi.connect.core.connectivity.TransportKind
import java.util.UUID

class LegacySessionObserver(
    val registry: SessionRegistry,
    private val differenceReporter: (String) -> Unit = {},
) {
    private var current: SessionId? = null

    fun observeStarted(id: UUID, transport: TransportKind, state: ConnectionState): SessionObservationResult {
        val sessionId = SessionId.parse(id.toString().replace("-", ""))
        current = sessionId
        return report(registry.observeStarted(sessionId, transport, state.toSessionState()))
    }

    fun observeState(state: ConnectionState): SessionObservationResult {
        val id = current ?: return report(SessionObservationResult(false, "No active legacy session."))
        if (state == ConnectionState.CONNECTED &&
            registry.snapshots.firstOrNull { it.id == id }?.state == SessionState.Connecting
        ) {
            report(registry.observeState(id, SessionState.Authenticating))
        }
        return report(registry.observeState(id, state.toSessionState()))
    }

    fun observeEnded(id: UUID): SessionObservationResult {
        val sessionId = SessionId.parse(id.toString().replace("-", ""))
        val result = registry.observeEnded(sessionId)
        if (current == sessionId) current = null
        return report(result)
    }

    fun observeClosedAll() {
        registry.observeClosedAll()
        current = null
    }

    private fun ConnectionState.toSessionState(): SessionState = when (this) {
        ConnectionState.IDLE, ConnectionState.DISCONNECTED -> SessionState.Closed
        ConnectionState.SEARCHING, ConnectionState.FOUND -> SessionState.Discovering
        ConnectionState.CONNECTING -> SessionState.Connecting
        ConnectionState.CONNECTED -> SessionState.Ready
        ConnectionState.STREAMING -> SessionState.Streaming
        ConnectionState.RECONNECTING -> SessionState.Reconnecting
        ConnectionState.ERROR -> SessionState.Failed
    }

    private fun report(result: SessionObservationResult): SessionObservationResult {
        if (!result.applied) differenceReporter("Legacy observation not applied: ${result.reason}")
        return result
    }
}

object SessionShadowComposition {
    fun createIfEnabled(readSetting: (String) -> String? = System::getProperty): LegacySessionObserver? =
        if (readSetting("modi.session.shadow") == "1") {
            LegacySessionObserver(SessionRegistry()) { message ->
                com.modi.connect.core.infrastructure.Log.w("SessionShadow", message)
            }
        } else null
}

File: windows/MoDi.Desktop/Connectivity/Sessions/LegacySessionObserver.cs
using System;
using System.Linq;
using MoDi.App.Contracts.Connectivity;
using MoDi.Core.Infrastructure;

namespace MoDi.Desktop.Connectivity.Sessions;

internal sealed class LegacySessionObserver
{
    private readonly SessionRegistry _registry;
    private SessionId? _current;

    internal LegacySessionObserver(SessionRegistry registry) => _registry = registry;

    internal SessionRegistry Registry => _registry;

    internal SessionObservationResult ObserveStarted(Guid id, TransportKind transport, ConnectionState state)
    {
        var sessionId = SessionId.Parse(id.ToString("N"));
        _current = sessionId;
        return Report(_registry.ObserveStarted(sessionId, transport, Map(state)));
    }

    internal SessionObservationResult ObserveState(ConnectionState state)
    {
        if (_current is not { } id)
            return Report(new(false, "No active legacy session."));

        if (state == ConnectionState.Connected &&
            _registry.Snapshots.FirstOrDefault(snapshot => snapshot.Id == id)?.State == SessionState.Connecting)
        {
            Report(_registry.ObserveState(id, SessionState.Authenticating));
        }
        return Report(_registry.ObserveState(id, Map(state)));
    }

    internal SessionObservationResult ObserveEnded(Guid id)
    {
        var sessionId = SessionId.Parse(id.ToString("N"));
        var result = _registry.ObserveEnded(sessionId);
        if (_current == sessionId) _current = null;
        return Report(result);
    }

    internal void ObserveClosedAll()
    {
        _registry.ObserveClosedAll();
        _current = null;
    }

    private static SessionState Map(ConnectionState state) => state switch
    {
        ConnectionState.Idle or ConnectionState.Disconnected => SessionState.Closed,
        ConnectionState.Searching or ConnectionState.Found => SessionState.Discovering,
        ConnectionState.Connecting => SessionState.Connecting,
        ConnectionState.Connected => SessionState.Ready,
        ConnectionState.Streaming => SessionState.Streaming,
        ConnectionState.Reconnecting => SessionState.Reconnecting,
        ConnectionState.Error => SessionState.Failed,
        _ => SessionState.Failed,
    };

    private static SessionObservationResult Report(SessionObservationResult result)
    {
        if (!result.Applied)
            Log.W("SessionShadow", $"Legacy observation not applied: {result.Reason}");
        return result;
    }
}

internal static class SessionShadowComposition
{
    internal static LegacySessionObserver? CreateIfEnabled(Func<string, string?>? readSetting = null)
    {
        readSetting ??= Environment.GetEnvironmentVariable;
        return string.Equals(readSetting("MODI_SESSION_SHADOW"), "1", StringComparison.Ordinal)
            ? new LegacySessionObserver(new SessionRegistry())
            : null;
    }
}
